examples/train-kernel-parameters/script.jl

```diff
@@ -0,0 +1,234 @@
+# # Train Kernel Parameters
+
+# Here we show a few ways to train (optimize) the kernel (hyper)parameters at the example of kernel-based regression using KernelFunctions.jl.
+
+# We load KernelFunctions and some other packages
+
+using KernelFunctions
+using LinearAlgebra
+using Distributions
+using Plots
+using BenchmarkTools
+using Flux
+using Flux: Optimise
+using Zygote
+using Random: seed!
+seed!(42);
+
+# ## Data Generation
+# We generate a toy dataset in 1 dimension:
+
+xmin, xmax = -3, 3  # Bounds of the data
+N = 50 # Number of samples
+x_train = rand(Uniform(xmin, xmax), N)  # sample the inputs
+σ = 0.1
+y_train = sinc.(x_train) + randn(N) * σ  # evaluate a function and add some noise
+x_test = range(xmin - 0.1, xmax + 0.1; length=300)
+nothing #hide
+
+# Plot the data
+
+scatter(x_train, y_train; label="data")
+plot!(x_test, sinc; label="true function")
+
+# ## Manual Approach
+# The first option is to rebuild the parametrized kernel from a vector of parameters 
+# in each evaluation of the cost fuction. This is similar to the approach taken in 
+# [Stheno.jl](https://github.com/JuliaGaussianProcesses/Stheno.jl).
+
+# To train the kernel parameters via ForwardDiff.jl,
+# we need to create a function creating a kernel from an array.
+# A simple way to ensure that the kernel parameters are positive
+# is to optimize over the logarithm of the parameters. 
+
+function kernelcall(θ)
+    return (exp(θ[1]) * SqExponentialKernel() + exp(θ[2]) * Matern32Kernel()) ∘
+           ScaleTransform(exp(θ[3]))
+end
+nothing #hide
+
+# From theory we know the prediction for a test set x given
+# the kernel parameters and normalization constant:
+
+function f(x, x_train, y_train, θ)
+    k = kernelcall(θ[1:3])
+    return kernelmatrix(k, x, x_train) *
+           ((kernelmatrix(k, x_train) + exp(θ[4]) * I) \ y_train)
+end
+nothing #hide
+
+# We look how the prediction looks like
+# with starting parameters [1.0, 1.0, 1.0, 1.0] we get :
+
+ŷ = f(x_test, x_train, y_train, log.(ones(4)))
+scatter(x_train, y_train; lab="data")
+plot!(x_test, sinc; lab="true function")
+plot!(x_test, ŷ; lab="prediction")
+
+# We define the loss based on the L2 norm both
+# for the loss and the regularization
+
+function loss(θ)
+    ŷ = f(x_train, x_train, y_train, θ)
+    return sum(abs2, y_train - ŷ) + exp(θ[4]) * norm(ŷ)
```

Comment:
Two questions:
a) why not
```suggestion
    return norm(y_train - ŷ) + exp(θ[4]) * norm(ŷ)
```
b) can you explain where the second term comes from? Just thinking about "parameter regularisation" might make one think it ought to be `norm(\theta)`.

Reply:
This part is from your original example, I am honestly not sure.